test/test_connectionpool.py

```diff
@@ -223,19 +224,66 @@ def test_max_connections(self):
 
             assert pool.num_connections == 1
 
-    def test_pool_edgecases(self):
+    def test_put_conn_when_pool_is_full_nonblocking(self):
+        """
+        If maxsize = n and we _put_conn n + 1 conns, the n + 1th conn will
+        get closed and will not get added to the pool.
+        """
         with HTTPConnectionPool(host="localhost", maxsize=1, block=False) as pool:
             conn1 = pool._get_conn()
-            conn2 = pool._get_conn()  # New because block=False
+            # pool.pool is empty because we popped the one None that pool.pool was initialized with
+            # but this pool._get_conn call will not raise EmptyPoolError because block is False
+            conn2 = pool._get_conn()
+
+            conn1.close = Mock()
+            conn2.close = Mock()
 
             pool._put_conn(conn1)
-            pool._put_conn(conn2)  # Should be discarded
+            pool._put_conn(conn2)
+
+            assert conn1.close.called is False
+            assert conn2.close.called is True
 
             assert conn1 == pool._get_conn()
             assert conn2 != pool._get_conn()
 
             assert pool.num_connections == 3
 
+    def test_put_conn_when_pool_is_full_blocking(self):
+        """
+        If maxsize = n and we _put_conn n + 1 conns, the n + 1th conn will
+        cause a FullPoolError.
+        """
+        with HTTPConnectionPool(host="localhost", maxsize=1, block=True) as pool:
+            conn1 = pool._get_conn()
+            conn2 = pool._new_conn()
+
+            conn1.close = Mock()
+            conn2.close = Mock()
+
+            pool._put_conn(conn1)
+            with pytest.raises(FullPoolError):
+                pool._put_conn(conn2)
+
+            assert conn1.close.called is False
+            assert conn2.close.called is True
+
+            assert conn1 == pool._get_conn()
```

Comment:
Should we do an `is` instead of `==` here?

Reply:
That totally makes sense to me, though other tests in this file use `==` for similar comparisons. Usually I would prefer to follow the convention of the rest of the file and perhaps leave a fixme.